kernel/cls/cls_macros/src/lib.rs

```diff
@@ -264,3 +273,237 @@ pub fn cpu_local(args: TokenStream, input: TokenStream) -> TokenStream {
+                        };
+
+                        unsafe {
+                            ::core::arch::asm!(
+                                "lea {from_cls_start}, [{cls}@TPOFF + {tls_size} + {cls_start_to_tls_start}]",
+                                from_cls_start = lateout(reg) from_cls_start,
+                                cls = sym #name,
+                                tls_size = in(reg) tls_size,
+                                cls_start_to_tls_start = in(reg) cls_start_to_tls_start,
+                                options(nomem, preserves_flags, nostack),
+                            )
+                        };
+                        let offset = (cls_size - from_cls_start).wrapping_neg();
+                        offset
+                    }
+                }
+            }
+            #[cfg(target_arch = "aarch64")]
+            {
+                extern "C" {
+                    static __THESEUS_TLS_SIZE: u8;
+                }
+
+                // TODO: Open an issue in rust repo? We aren't actually doing anything unsafe.
+                // SAFETY: We don't access the extern static.
+                let tls_size = unsafe { ::core::ptr::addr_of!(__THESEUS_TLS_SIZE) } as u64;
+
+                // Unlike x64, AArch64 doesn't have different branches for statically linked, and
+                // dynamically linked variables. This is because on AArch64, there are no negative
+                // offset shenanigans; a TLS data image looks like
+                // ```
+                // fs
+                // V
+                // +-----------------------+------------------------+
+                // | statically linked TLS | dynamically linked TLS |
+                // +-----------------------+------------------------+
+                // ```
+                //
+                // Unlike x64, the `.cls` section is located after `.tls` in a binary and so the
+                // linker thinks the data image looks like
+                // ```
+                // fs
+                // V
+                // +-----------------------+-----+-----------------------+------------------------+
+                // | statically linked TLS | 000 | statically linked CLS | dynamically linked TLS |
+                // +-----------------------+-----+-----------------------+------------------------+
+                // ```
+                // where `000` are padding bytes to align the start of the statically linked CLS to
+                // a page boundary.
+                //
+                // Hence, CLS symbols have an offset that is incorrect by
+                // `tls_size.next_multiple_of(0x1000)`, or 0 if there is not TLS section. So, when
+                // loading CLS symbols on AArch64, `mod_mgmt` simply sets `__THESEUS_TLS_SIZE` to 0.
+
+                let tls_start_to_cls_start: u64 = if tls_size == 0 {
+                    0
+                } else {
+                    // TODO: Use `next_multiple_of` when stabilised.
```

Comment:
It would require the implementor crate to add `#![feature(int_roundings)]` to the crate root for no real gain.

Rust does manage to make it branchless but yea probably best to make it explicit.